File: lib/composite/directory.rb
class Directory < Entry
  def initialize(name)
    @name = name
    @directory = []
  end

  def get_name
    @name
  end

  def get_size
    @directory.sum { |entry| entry.get_size }
  end

  def print_list(prefix = "")
    puts("#{prefix}/#{self}")
    @directory.each do |entry|
      entry.print_list("#{prefix}/#{@name}")
    end
  end

  def add(entry)
    @directory << entry
    self
  end
end

File: lib/visitor/directory.ts
import { Entry } from "./entry";
import { Visitor } from "./visitor";

export class Directory extends Entry implements Iterable<Entry> {
  private directory: Entry[] = [];

  constructor(private name: string) {
    super();
  }

  getName(): string {
    return this.name;
  }

  getSize(): number {
    return this.directory.reduce((sum, e) => sum + e.getSize(), 0);
  }

  add(entry: Entry): Directory {
    this.directory.push(entry);
    return this;
  }

  [Symbol.iterator](): Iterator<Entry> {
    return this.directory[Symbol.iterator]();
  }

  accept(v: Visitor): void {
    v.visitDirectory(this);
  }
}
